internal/apis/certmanager/types_issuer.go

```diff
@@ -187,6 +187,27 @@ type VaultIssuer struct {
 	// parameter is ignored for plain HTTP protocol connection. If not set the
 	// system root certificates are used to validate the TLS connection.
 	CABundle []byte
+
+	// Like CABundle but loads the CA bundle from a Kubernetes Secret.
+	// CABundle will take precedence if also set.
```

Comment:
```suggestion
	// CABundleConfigMapRef is a reference to a ConfigMap which contains the CABundle which will be used when connecting to Vault when using HTTPS. Cannot be used in conjunction with CABundle. Is neither CABundle nor CABundleConfigMapRef are defined, the cert-manager controller system root certificates are used to the validate the TLS connection.
```

We should also update the `CABundle` comment with the above constraints as well.

Reply:
Is there a particular reason to use a Secret here, rather than using a ConfigMap? In Kubernetes it is typically the case the CA bundles are stored in ConfigMap resources. (see [Kubernetes CA](https://kubernetes.io/docs/tasks/tls/managing-tls-in-a-cluster/#trusting-tls-in-a-cluster), [cert-manager/trust](https://cert-manager.io/docs/projects/trust/#usage), [istio](https://github.com/cert-manager/istio-csr/blob/e995c3561738c423adcf76ae2ff2a6482c682f72/pkg/controller/configmap.go#L42), etc.). By making this a ConfigMap, users would have a higher chance to be able to automate this ConfigMap with other tools.


We likely want to create a new meta type `ConfigMapKeySelector`, which is a carbon copy of [`SecretKeySelector`](https://github.com/cert-manager/cert-manager/blob/6947696b12428a1f9e5823c2e69de84b8dd08c1c/pkg/apis/meta/v1/types.go#L65), but is named and commented to make it clear that it acts on ConfigMaps, rather than Secrets.

Reply:
We should also add validation to the [webhook](https://github.com/cert-manager/cert-manager/blob/6947696b12428a1f9e5823c2e69de84b8dd08c1c/internal/apis/certmanager/validation/issuer.go#L227) to ensure that both `CABundle` and `CABundleConfigMapRef` cannot both be defined.

Reply:
[Bank Vaults](https://github.com/banzaicloud/bank-vaults) uses am opaque secret for storing the bundle and that's currently my primary use-case.

Reply:
It is not so much having the ability to add more than one certificate to the bundle at a time - that's totally fine and _should_ be supported - it's the fact we have two fields here which are doing the same thing.. I don't have super strong opinions, but would prefer we didn't combine them.

---

Is there no way to configure it to write to a ConfigMap instead, or maybe using another project like [`cert-manager`](https://github.com/cert-manager/trust) to distribute the bundle? Unfortunate to use Secrets, when the current precedent is to use ConfigMaps.

Reply:
I'd argue that there is hardly any demand for loading the bundle via a Configmap. The caBundle field was added in 2018 and as far as I can tell nobody has asked for a Configmap configuration option yet. I was only able to find one comment that asks for a configuration via a secret https://github.com/cert-manager/cert-manager/issues/3363#issuecomment-786159074
The only use-case that I can think of where you'd even need a dynamic CA bundle configuration approach is the one I have: Vault provisioning via an operator while using a GitOps workflow. I don't have experience with any other Vault operators than Bank Vaults but I'd assume that they also store the Vault key bundle in a single secret from which you then might want to read the CA cert. Of course, giving cert-manager access to a secret that also contains the Vault private key might not be desirable in most cases but that should be up to the users and their specific threat model. At least in the case of Bank Vaults, it removes the private key from the secret before automatically distributing it to other namespaces.

Reply:
Also, a secret would play nicely with [this PR](https://github.com/cert-manager/cert-manager/pull/4330) as the Secret for the client auth would most likely also contain the ca.crt, i.e. the same Secret could then be referenced both for the client auth and the CA bundle.

Reply:
It's an aside, but the `ca.crt` key in resulting client TLS Secrets was a bit of a mistake.. and shouldn't be used for trust distribution. You can read more on the topic on the [concepts page](https://cert-manager.io/docs/concepts/certificate/).